src/cppsim/simulator.cpp

```diff
@@ -9,12 +9,17 @@
 QuantumCircuitSimulator::QuantumCircuitSimulator(
     QuantumCircuit* circuit, QuantumStateBase* initial_state)
     : _circuit(circuit), _state(initial_state), _buffer(NULL) {
-    if (_state == NULL) _state = new QuantumState(this->_circuit->qubit_count);
+    if (_state == NULL) {
+        _state = new QuantumState(this->_circuit->qubit_count);
+        _own_state = true;
+    }
+    // _circuitはQuantumCircuitSimulatorを継承しているParametricQuantumCircuitSimulatorが
+    // circuitのポインタを共有しておりtestで最適化した回路を使用するので、
+    // 渡されたcircuitをそのまま使用する必要がある。
 };
 
 QuantumCircuitSimulator::~QuantumCircuitSimulator() {
-    if (_circuit != NULL) delete _circuit;
-    if (_state != NULL) delete _state;
+    if (_own_state) delete _state;
```

Comment:
`_state` points to the `QuantumStateBase` which is not owned by this class, so this destructor should not  `delete` the `_state`. This should be deleted at the scope where the state is initialized.

Reply:
I had the wrong code.
https://github.com/qulacs/qulacs/blob/fix-344-simulator-abort/src/cppsim/simulator.cpp#L12
The check here was intended to ensure that the initial_state was NULL.
I'll fix it.

Reply:
Fixed checking of initialization parameters.
if initial _ state is null, new _state.
And delete destructor.